enhancements/update/reduced_reboots.md

```diff
@@ -0,0 +1,328 @@
+whenever possible.
+
+#### OTA - Implement Downstream Paused Worker Pool Upgrade Tests
+
+In parallel with efforts to revamp upstream version skew testing we must also
+implement downstream version skew testing which includes any additional tests
+required for OpenShift specific implementation details.
+
+We will achieve this by delivering upgrade jobs which pause the Worker MachineConfigPool
+then upgrade from 4.x to 4.x+1 to 4.x+2. We will run conformance tests from 4.x
+after upgrading to 4.x+1 in order to ensure that we continue to provide a baseline
+feature set, then again after upgrading to 4.x+2, finally after unpausing the
+Worker MCP we will run 4.x+2 tests.
+
+Given the complexity of these test jobs we should expect that they may take
+longer than the current four hour limit for test jobs. Rather than compromising
+on test completeness we will seek to extend test duration limits or find other
+ways to meet these testing demands.
+
+#### Teams with Host Components - Allow N-2 Host Component Version Skew
+
+All teams which own components that directly interface with or ship host based
+components will need to ensure that they're broadening their compatibility to
+allow for N-2 version skew between host and cluster scoped resources.
+
+This would include for example the SDN DaemonSets in 4.10 remaining compatible
+with OVS and any other host components in 4.10, 4.9, and 4.8. On a case by case
+basis teams should decide whether it makes more sense to maintain a broader
+compatibility matrix or that N-1 bits and MachineConfig are backported to N-2 and
+upgrade graph is amended with these new minimum version requirements.
+
+For instance, if 4.9.12 is the minimum version for 4.9 to 4.10 upgrades we'd
+ensure that the next 4.8.z shipping after 4.9.12 has RHCOS bits and MachineConfig
+which offer parity with 4.9.12 so that it's not required that we reboot into
+4.9.12. If teams choose to pursue this option they will need to continue to ensure
+that 4.7 to 4.8.z and 4.8.z-n to 4.8.z upgrades continue to work as well.
+
+Teams which believe this is not achievable or the level of effort is extremely
+high should document those findings.
+
+Thus far RHCOS, Node, MCO, SDN, Containers, and PSAP teams are known to fall into
+this group of teams which have components coupled to the host.
+
+#### MCO - Widen Node Constraints to allow for N-2
+
+Building upon the EUS-to-EUS upgrade MVP work(https://github.com/openshift/enhancements/blob/master/enhancements/update/eus-upgrades-mvp.md#mco---enforce-openshifts-defined-host-component-version-skew-policies)
+to allow MCO to enforce host constraints we will broaden those constraints to
+enable host component version skew.
+
+Admins who choose to would then be able to skip a host reboot by following this
+pattern:
+
+1. Starting with a 4.8 cluster, pause Worker MachineConfigPool
+1. Upgrade to 4.9
+1. Upgrade to 4.10
+1. Unpause Worker MachineConfigPool
```

Comment:
What about custom MachineConfigPools?

Reply:
One example would be e.g. real-time kernel workers that are in a separate MCP with additional MachineConfigs

Reply:
All pools should inherit from `worker`, so if users have custom pools these would be updated when MCO generates a new rendered config for worker on every step. Custom pools needs to be paused as well

Reply:
Do those inherit from Worker pool? This does make me wonder if we need to define our policies on a per pool basis, especially if we're considering that control plane pool must be updated.

@darkmuggle another point to consider here.

Reply:
Ok just for clarification, if I pause a custom MCP this property will not be backpropagated to the parent worker MCP. MCO will rollout the upgrade on the worker MCP and since my custom MCP is paused it will wait until availability.

Reply:
@zvonkok reference for that? I'd be interested in the set of annotations that would provide.

Reply:
@sdodson This is needed to support https://github.com/openshift/enhancements/pull/357, https://issues.redhat.com/browse/GRPA-2713

Reply:
@sdodson This was a question. :) Forgot the question mark.

Reply:
> Custom pools needs to be paused as well

I don't think any pools _need_ to be paused.  Folks interested in reducing reboots can just pause the pools in which they want to reduce reboots, for the short time it takes to make a few consecutive update hops.  I don't understand how pool-pausing interacts between the `worker` pool and its custom decedents; maybe the MCO folks can speak to that.

Reply:
Users could have self-defined constraints that are not yet applied or implemented in a piece of software. Either an operator is preventing the upgrade or an admin wishes to do some "day 2" customizations before upgrading, knowing that an upgrade could break something.

Reply:
What about workloads that are running _longer_ (days, weeks) that cannot be checkpointed, wouldn't this also be a reason to pause an MCP? PodDisruptionBudgets with minAvailable=100% would prevent draining to finish but the Node would already be cordoned. The workload may need additional Pods to be scheduled on the node to finish the task.

Reply:
> PodDisruptionBudgets with minAvailable=100% would prevent draining to finish but the Node would already be cordoned.

There is a balance between wasting resources by not scheduling work on the node while you wait for the slow pod to wrap up, and getting stuck in an endless drain because the new work you let onto the node ends up also being guarded by a PDB or other mechanism and taking longer than the initial slow pod.  Currently we weight in favor of quickest-possible-drain at the expense of underutilizing the available capacity.  Folks who want to minimize wasted capacity can set `maxUpgradeable: 1` (the default) on their MachineConfigPools, although the net effect is just spreading the waste out over a longer wall-clock duration as the nodes roll one after the other.  That spreading may still be useful if the overall pool doesn't have much extra capacity to spare.  And if folks feel really strongly, they can arrange to fill each node with workloads that will all wrap up around the same point in time, or twiddle tolerations to allow faster work onto the cordoned node until the slow work gets closer to wrapping up.

> The workload may need additional Pods to be scheduled on the node to finish the task.

It's up to the actor managing that workload to either say "you know what, we should clear this slow work off so that poor node can update" or "I am unwilling to abandon this slow work, so I'm going to set sufficient tolerations and PDB guards on these new Pods so they can go join the slow worker on the cordoned node and help it push through to completion".  I don't think that's something generic OpenShift tooling that doesn't understand the details of your workload can help you with out of the box.  In both of these cases, pausing the pool is one way to reduce the likelihood of cordons and node reboots.  But MachineConfigs are not the only reason your nodes might get cordoned or rolled.  I think it's better for folks who have really expensive, slow work to write their own really smart controllers who can shepherd the work through a generic Kube-node environment instead of using the coarse pool-pausing knob to try to orchestrate the dance they want to see.

Reply:
@wking Right, we do not have to lengthen this. Either you want to update or you do not want to. If you want to update then make a plan on how to fit the upgrade procedure into your daily/weekly business.  We can offer the tools and mechanics but "you" should do "your" homework.